File: CPP/NumberOfSubarraysWithBoundedMaximum.cpp
#include <iostream>
#include <vector>

class Solution {
public:
  int numSubarrayBoundedMax(std::vector<int> &nums, int left, int right) {
    return count(nums, right) - count(nums, left - 1);
  }

private:
  int count(const std::vector<int> &nums, int bound) {
    int total = 0, length = 0;
    for (int x : nums) {
      length = (x <= bound) ? length + 1 : 0;
      total += length;
    }
    return total;
  }
};

int main() {
  Solution s;
  std::vector<int> a = {2, 1, 4, 3}, b = {2, 9, 2, 5, 6}, c = {1, 1, 1};
  std::cout << s.numSubarrayBoundedMax(a, 2, 3) << "\n"; // 3
  std::cout << s.numSubarrayBoundedMax(b, 2, 8) << "\n"; // 7
  std::cout << s.numSubarrayBoundedMax(c, 1, 1) << "\n"; // 6
  return 0;
}

File: Python/NumberOfSubarraysWithBoundedMaximum.py
from typing import List


class Solution:
    def numSubarrayBoundedMax(self, nums: List[int], left: int, right: int) -> int:
        # count(b) = # subarrays whose every element is <= b (a running length of consecutive
        # in-bound elements contributes len subarrays ending here). Subarrays with max in
        # [left, right] = count(right) - count(left - 1). O(n).
        def count(bound: int) -> int:
            total = 0
            length = 0
            for x in nums:
                length = length + 1 if x <= bound else 0
                total += length
            return total

        return count(right) - count(left - 1)


if __name__ == "__main__":
    s = Solution()
    print(s.numSubarrayBoundedMax([2, 1, 4, 3], 2, 3))  # 3
    print(s.numSubarrayBoundedMax([2, 9, 2, 5, 6], 2, 8))  # 7
    print(s.numSubarrayBoundedMax([1, 1, 1], 1, 1))  # 6
